include/oneapi/dpl/pstl/iterator_impl.h

```diff
@@ -401,7 +401,7 @@ class transform_iterator
 {
   private:
     _Iter __my_it_;
-    const _UnaryFunc __my_unary_func_;
+    _UnaryFunc __my_unary_func_;
```

Comment:
What do you think about adding a static_assert to enforce the callable to be non-mutable?

Reply:
We already require that the call operator is const with the `const` in the dereference operator of `transform_iterator`.
`reference operator*() const { return __my_unary_func_(*__my_it_); }` 

Are you meaning to check with a `static_assert` for the availability of an appropriate const call operator of the proper signature?  I'll have to experiment with how this can be done (without sacrificing our default ctor).

I don't think there is a way to add a compile time check that the functor has no `mutable` fields.

Reply:
Looking at `oneapi::dpl::__internal::__is_const_callable_object_v`, it does not work properly whenever there are multiple overloads of `operator()` (including if there is a templated `operator()`).  I don't want to use this here, and we may want to examine it's usage in `replace_copy_if`. Adding a more robust version of this trait for arbitrary signatures is probably beyond the scope of this PR.

However, I've added a check that a const version of the functor is callable with a specific argument type, and added a `static_assert` to the structure definition.  This could probably be expanded to arbitrary argument list rather than a single argument with more effort.

The benefit this provides would be a better error message for the user. It also checks upon instantiation of the type, and doesn't require attempted dereference of the `transform_iterator` within the code.

What do you think?

Reply:
> Adding a more robust version of this trait for arbitrary signatures is probably beyond the scope of this PR.

I guess it is called `std::is_invocable_v` :) But yes. switching to that is beyond the scope of this PR.

Reply:
`is_invocable_v` is probably what we want to be checking within `replace_copy_if` with a specific argument type.  However, it is probably not well suited to implement the general helpers `__is_callable_object`, `__is_const_callable_object` as those aren't taking into account the arguments we want to pass in (those can probably be removed though).  

I'll create an issue for this.

Reply:
> oneapi::dpl::__internal::__is_const_callable_object_v, does not work properly

I agree. It's broken. I forgot how exactly broken, but broken